plugins/balloonpanel/plugin.js

```diff
@@ -363,6 +408,34 @@
 			};
 
 			return function( element, options ) {
+				if ( element instanceof CKEDITOR.dom.selection ) {
+					var ranges = element.getRanges(),
+						rectList = ranges[ ranges.length - 1 ].getClientRects( true );
+
+					// If selection is fake we have selected widget or table.
+					if ( element.isFake ) {
```

Comment:
After the quick test in Chrome, it seems that this step could be omitted for widget ranges at all, as the new `rectList` after the loop is the same as the `rectList` before it. We have https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_selection.html#method-isInTable method to check if selection is in table.

IMO it would be even better to refactor this whole code to something like this:

```javascript
if ( element.isFake && element.isInTable() ) {
	// loop and prepare rects
} else {
	// just getClientRects of appropriate range
}
```

After this refactor rectlist will be calculated only once in every situation.